src/change_stream.ts

```diff
@@ -107,94 +108,220 @@ export interface PipeOptions {
+export interface ChangeStreamFullNameSpace {
+  /** The namespace */
+  ns: { db: string; coll: string };
+}
+
+/** @public */
+export interface ChangeStreamDocumentKey<TSchema extends Document = Document> {
   /**
-   * The id functions as an opaque token for use when resuming an interrupted
-   * change stream.
+   * For unsharded collections this contains a single field `_id`.
+   * For sharded collections, this will contain all the components of the shard key
    */
-  _id: InferIdType<TSchema>;
+  documentKey: { _id: InferIdType<TSchema>; [shardKey: string]: any };
+}
 
+/** @public */
+export interface ChangeStreamDocumentCommon {
   /**
-   * Describes the type of operation represented in this change notification.
+   * The id functions as an opaque token for use when resuming an interrupted
+   * change stream.
    */
-  operationType:
-    | 'insert'
-    | 'update'
-    | 'replace'
-    | 'delete'
-    | 'invalidate'
-    | 'drop'
-    | 'dropDatabase'
-    | 'rename';
-
+  _id: ResumeToken;
   /**
-   * Contains two fields: “db” and “coll” containing the database and
-   * collection name in which the change happened.
+   * The timestamp from the oplog entry associated with the event.
+   * For events that happened as part of a multi-document transaction, the associated change stream
+   * notifications will have the same clusterTime value, namely the time when the transaction was committed.
+   * On a sharded cluster, events that occur on different shards can have the same clusterTime but be
+   * associated with different transactions or even not be associated with any transaction.
+   * To identify events for a single transaction, you can use the combination of lsid and txnNumber in the change stream event document.
    */
-  ns: { db: string; coll: string };
+  clusterTime?: Timestamp;
 
   /**
-   * Only present for ops of type ‘insert’, ‘update’, ‘replace’, and
-   * ‘delete’.
-   *
-   * For unsharded collections this contains a single field, _id, with the
-   * value of the _id of the document updated.  For sharded collections,
-   * this will contain all the components of the shard key in order,
-   * followed by the _id if the _id isn’t part of the shard key.
+   * The transaction number.
+   * Only present if the operation is part of a multi-document transaction.
    */
-  documentKey?: { _id: InferIdType<TSchema> };
+  txnNumber?: number | Long;
 
   /**
-   * Only present for ops of type ‘update’.
-   *
-   * Contains a description of updated and removed fields in this
-   * operation.
+   * The identifier for the session associated with the transaction.
+   * Only present if the operation is part of a multi-document transaction.
    */
-  updateDescription?: UpdateDescription<TSchema>;
+  lsid?: ServerSessionId;
+}
 
+/**
+ * @public
+ * @see https://www.mongodb.com/docs/manual/reference/change-events/#insert-event
+ */
+export interface InsertChangeStreamDocument<TSchema extends Document = Document>
```

Comment:
nit: I think `ChangeStream<Command>Result` might make more sense than `<Command>ChangeStreamDocument` as a general format for these interfaces but I don't feel too strongly.  Thoughts?

Reply:
Renamed them all, I'm still using Document at the end because I don't want to change the name of the `ChangeStreamDocument` type and we should keep these consistent, but reordering the indicator makes sense.